### ISSUE TITLE
pihole won't start with latest on 8/22/22

### ISSUE THREAD

**Issue by dsm1212** [pi-hole/docker-pi-hole#1174]
I pulled the latest docker today 8/22/22. I pull updates on all images every friday but skipped this last friday. So something that changed in the last 10 days. I'm not getting much in the log. Just shows that a service didn't start and so is shutting down. It mentions checking some /run container log, but I can't get detached for more than a second. Sorry not much to go on but my network is toast when this happens. I've reverted to image 52916ce4626f and everything works. What can I gather? Is there a file I can map to get the service startup logging?

**Comment by dsm1212** [comment]
My compose info (I'm using portainer):

  pihole:
    container_name: pihole
    image: pihole/pihole:best
    network_mode: host
    environment:
      TZ: 'America/New_York'
      INTERFACE: bond0
      WEBPASSWORD: <redacted>
      WEB_PORT: 4080
      PIHOLE_DNS_: '127.0.0.1#5335;;0:0:0:0:0:0:0:1#5335'
      FTLCONF_REPLY_ADDR4: '192.168.1.201'
      FTLCONF_REPLY_ADDR6: <redacted>
      REV_SERVER: 'true'
      REV_SERVER_TARGET: '192.168.1.1'
      REV_SERVER_DOMAIN: <redacted>
      REV_SERVER_CIDR: '192.168.1.0/24'
      ServerIP: '192.168.1.201'
      ServerIPv6: <redacted>
      VIRTUAL_HOST: pihole.glazed.duckdns.org
      TEMPERATUREUNIT: 'f'
      WEB_UID: 1021
      WEB_GID: 100
      PIHOLE_UID: 1021
      PIHOLE_GID: 100
    volumes:
      - '/apps/pihole/:/etc/pihole/'
      - '/apps/pihole/dnsmask.d/:/etc/dnsmasq.d/'
    cap_add:
      - NET_ADMIN
      - SYS_PTRACE
    restart: unless-stopped

**Comment by dsm1212** [comment]
Ok, here is log: 

`s6-rc: info: service cron: starting
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service cron successfully started
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service _startup: starting
Starting crond
ServerIP is deprecated. Converting to FTLCONF_REPLY_ADDR4
ServerIPv6 is deprecated. Converting to FTLCONF_REPLY_ADDR6
fix-attrs: info: applying /etc/fix-attrs.d/01-resolver-resolv
fix-attrs: warning: fix-attrs is deprecated, please fix volume permissions in your container manager instead
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
cont-init: info: running /etc/cont-init.d/05-changer-uid-gid.sh
cont-init: info: /etc/cont-init.d/05-changer-uid-gid.sh exited 0
 ::: Starting docker specific checks & setup for docker pihole/pihole
s6-rc: info: service legacy-cont-init successfully started
  [i] Installing configs from /etc/.pihole...
  [i] Existing dnsmasq.conf found... it is not a Pi-hole file, leaving alone!
  [i] Installing /etc/dnsmasq.d/01-pihole.conf...
  [✓] Installed /etc/dnsmasq.d/01-pihole.conf
  [i] Installing /etc/.pihole/advanced/06-rfc6761.conf...
  [✓] Installed /etc/dnsmasq.d/06-rfc6761.conf
  [i] Installing latest logrotate script...
	[i] Existing logrotate file found. No changes made.
Custom WEB_PORT set to 4080
INFO: Without proper router DNAT forwarding to 192.168.1.201:4080, you may not get any blocked websites on ads
::: Assigning password defined by Environment Variable
  [✓] New password set
  [✓] Set temperature unit to F
Added ENV to php:
			"TZ" => "America/New_York",
			"PIHOLE_DOCKER_TAG" => "2022.08",
			"PHP_ERROR_LOG" => "/var/log/lighttpd/error-pihole.log",
			"CORS_HOSTS" => "",
			"VIRTUAL_HOST" => "<redacted>",
Using IPv4 and IPv6
::: Preexisting ad list /etc/pihole/adlists.list detected ((exiting setup_blocklists early))
https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
Setting DNS servers based on PIHOLE_DNS_ variable
s6-rc: fatal: timed out
/run/s6/basedir/scripts/rc.init: warning: s6-rc failed to properly bring all the services up! Check your logs (in /run/uncaught-logs/current if you have in-container logging) for more information.
prog: fatal: stopping the container.
s6-sudoc: fatal: unable to get exit status from server: Operation timed out
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service cron: stopping`

**Comment by rdwebdesign** [comment]
Can you please try to create a new container, with a different name (example: `piholetest`) and without volumes?

This will test how the image behave in your environment, without any additional config.

**Comment by dsm1212** [comment]
Yeah I tried that and it starts fine. I looked over my compose and commented out the serverip lines which are deprecated and I removed a extra semicolon in the PIHOLE_DNS_ entry. After doing those two things it came up. I think maybe it was the extra semicolon. Back when I started I did that so it would fill the first and third entries I the settings window since the 3rd one is for ipv6, but probably was not necessary and for some reason it caused a problem now.

**Comment by devopsleigh** [comment]
I had the same error `s6-rc: fatal: timed out
/run/s6/basedir/scripts/rc.init: warning: s6-rc failed to properly bring all the services up! Check your logs (in /run/uncaught-logs/current if you have in-container logging) for more information.`

Starting without volumes works just fine @rdwebdesign, however adding an empty path to the volume will cause it to fail despite populating it with files.

I reverted to 2022.07.1 and the container starts normally.

**Comment by xtreemrage** [comment]
@devopsleigh 
Do you use an own dockerfile? I did and my entry point was `/s6-init`, when I changed it to `/init` it started to work for me.

**Comment by devopsleigh** [comment]
> @devopsleigh
> Do you use an own dockerfile? I did and my entry point was `/s6-init`, when I changed it to `/init` it started to work for me.

No, I have been using pihole/latest until I have issues and then I pin to a version. I'm not that deep in Docker knowledge to understand S6.

**Comment by sbernesto** [comment]
> @devopsleigh Do you use an own dockerfile? I did and my entry point was `/s6-init`, when I changed it to `/init` it started to work for me.

I still have the same error and I corrected with this steps

**Comment by PromoFaux** [comment]
https://github.com/pi-hole/docker-pi-hole#upgrading--reconfiguring

**Comment by dsm1212** [comment]
Just guessing, but I think the issue is that the s6 change is less tolerant of messages or return codes from these startup scripts. The double semicolon I had was causing a "warning" message to come out. I had to start a container using sh as the entry point and then run the init script steps by hand to see it. Maybe there are other "warning" cases like this which are now actually errors.

**Comment by PromoFaux** [comment]
I  probably should have been a little more verbose in my response. We often see that users on portioner are not necessarily updating their containers "correctly". One needs to completely destroy the container and recreate a fresh one, else things can get stuck in the environment from previous iterations.

Personally I have defined mine in Portainer as a "stack" (which is basically docker compose), and after pulling the latest image I just press:

![image](https://user-images.githubusercontent.com/1998970/186230015-b318e1c3-6a68-4b75-a0b3-13f883383a53.png)

then

![image](https://user-images.githubusercontent.com/1998970/186230124-1e4597b0-7282-496b-997e-2725e74bf77a.png)

Which seems to do the job.

On the command line I would have run:

```
docker-compose pull
docker-compose down
docker-compose up -d
```

For the same thing

**Comment by rdwebdesign** [comment]
In Portainer, there is a button to duplicate a container, but a lot of users mixes the ideas between Portainer's "Duplicate" and create a new container.

This button is named "Duplicate/Edit" because it actually use the same config already in use and allows the user to make a few adjustments.
But almost every config is kept (including the entrypoint, ENV VARs added by the image, container labels, capability settings).

This feature is good when you want to test multiple similar configurations for the same image, changing just a few items.

If you use a previous config with a new image, it will probably fail.

There is a warning on the top of the page:
![image](https://user-images.githubusercontent.com/1385443/186250673-f1524b5f-5be7-4cb2-be72-9fe6c9d7b0ae.png)

**Comment by devopsleigh** [comment]
Adding my 2c...

I always use `docker compose up -d --force-recreate` and I'm seeing the same behaviour.

**Comment by dsm1212** [comment]
> In Portainer, there is a button to duplicate a container, but a lot of users mixes the ideas between Portainer's "Duplicate" and create a new container.

Ah, I see, good to know :-). I think the issue as I said earlier is that the new s6 is not tolerant of some "warning" level things that were ignored previously. Both cases here are produced with a minor issue in the env values that were previously benign. The third case here was due to a custom docker file not updated so I would ignore that one.

**Comment by dsm1212** [comment]
Rather than try to ignore errors, maybe you could find a way to reflect the s6 startup logging to the container output? Or cat it there on error maybe. The problem is that the error message we get is not actionable since the container never gets started.

**Comment by svenblum** [comment]
I know we shouldn't update pihole automatically. But it worked for the last yearS. And now your last 2 releases are broken - only a local login and local start of the container fixes it. But it is very annoying! Please DO FIX IT before you release another update, please!!!

**Comment by PromoFaux** [comment]
> I know we shouldn't update pihole automatically. But it worked for the last yearS

And yet, things can and do break. We're a team of volunteers doing as much as we have the time to, we can't test for every single combination of hardware/esoteric container orchestration system out there. This is why we are _explicit_ in our suggestion not to auto update your containers. If you want to ignore that advice and go ahead with automatic update, well that's on you :)

https://github.com/pi-hole/docker-pi-hole#note-on-watchtower

TL;DR - Read the release notes, set aside some time to do the upgrade, the do the upgrade manually.

It's also not like we didn't ask for help testing it a month before we released:

https://twitter.com/The_Pi_Hole/status/1550530329403764737?cxt=HHwWgsClmdPHy4QrAAAA

`latest` / `2022.08.2` should now have fixes for issues that have been reported since releasing `2022.08`. Thank you for those who filed bug reports

**Comment by devopsleigh** [comment]
Thanks @PromoFaux, that was a fast turnaround

**Comment by dsm1212** [comment]
I looked over the changes though and I'm not sure if these issues were addressed. Were they caused by the location change of /s6-init? They seem more like the update of s6 resulted in different error handling? I could be wrong, I just don't see how these changes 2022.08.2 have anything to do with the flawed environment settings for volumes and dns servers.

**Comment by PromoFaux** [comment]
Sorry, there were a few different things that were similar that I may have conflated a few things.

Your issue was caused by configuration with the `;;`, and looking at it I don't think it was the `s6` change that caused it, but rather a change in the code:

`2022.07.1`:

https://github.com/pi-hole/docker-pi-hole/blob/9fd7719638f4984725cd9cafe9063724cbb5bbbd/start.sh#L122

I guess I refactored this line when moving it to `bash_functions.sh` (`2022.08` and higher) ([reference PR](https://github.com/pi-hole/docker-pi-hole/pull/1141)):

https://github.com/pi-hole/docker-pi-hole/blob/9039a73272e91d811ccda3316d5c5ddb751b7e91/src/scripts/bash_functions.sh#L269

>Back when I started I did that so it would fill the first and third entries I the settings window since the 3rd one is for ipv6, but probably was not necessary and for some reason it caused a problem now.

Re-reading your report above, yeah, not necessary - but I see what you were _trying_ to do. That all said - you don't actually need to have an upstream ipv6 resolver - as an IPv4 resolver will still return IPv6 answers to queries!

With the new code, using two `;;` causes some odd output:

```
adam@adam-pc:~/playground/test$ ./test.sh 
Setting DNS servers based on PIHOLE_DNS_ variable
PIHOLE_DNS_1 1.1.1.1
Resolved  from PIHOLE_DNS_ as: h.root-servers.net.
Invalid entry detected in PIHOLE_DNS_: 
PIHOLE_DNS_2 8.8.8.8
```

And I believe it is the thrown error in the middle there that was causing the s6 scripts to crap out.

With the old code. no errror:

```
adam@adam-pc:~/playground/test$ ./test.sh 
Setting DNS servers based on PIHOLE_DNS_ variable
PIHOLE_DNS_1 1.1.1.1
PIHOLE_DNS_2 8.8.8.8
```

I will look into more versbose warning/error logging in the future, but for now at least we know that it was down to a misconfigured value!

The reason for the change on that line, for the curious, was https://www.shellcheck.net/wiki/SC2206
